src/ImageSharp/Common/Helpers/SimdUtils.HwIntrinsics.cs

```diff
@@ -140,154 +165,168 @@ private static Vector256<byte> ShuffleMaskShiftAlpha() => Vector256.Create((byte
 
                 if (sourceCount > 0)
                 {
                     Shuffle4Slice3(
                         source[..sourceCount],
-                        dest[..destCount],
+                        destination[..destinationCount],
                         control);
 
                     source = source[sourceCount..];
-                    dest = dest[destCount..];
+                    destination = destination[destinationCount..];
                 }
             }
         }
 
         [MethodImpl(InliningOptions.ShortMethod)]
         private static void Shuffle4(
             ReadOnlySpan<float> source,
-            Span<float> dest,
+            Span<float> destination,
             [ConstantExpected] byte control)
         {
-            if (Avx.IsSupported)
+            if (Vector512.IsHardwareAccelerated && Vector512Utilities.SupportsShuffleFloat)
             {
-                ref Vector256<float> sourceBase =
-                    ref Unsafe.As<float, Vector256<float>>(ref MemoryMarshal.GetReference(source));
+                ref Vector512<float> sourceBase = ref Unsafe.As<float, Vector512<float>>(ref MemoryMarshal.GetReference(source));
+                ref Vector512<float> destinationBase = ref Unsafe.As<float, Vector512<float>>(ref MemoryMarshal.GetReference(destination));
 
-                ref Vector256<float> destBase =
-                    ref Unsafe.As<float, Vector256<float>>(ref MemoryMarshal.GetReference(dest));
+                nuint n = (uint)destination.Length / (uint)Vector512<float>.Count;
+                nuint m = Numerics.Modulo4(n);
+                nuint u = n - m;
 
-                nint n = (nint)dest.Vector256Count<float>();
-                nint m = Numerics.Modulo4(n);
-                nint u = n - m;
-
-                for (nint i = 0; i < u; i += 4)
+                for (nuint i = 0; i < u; i += 4)
                 {
-                    ref Vector256<float> vd0 = ref Unsafe.Add(ref destBase, i);
-                    ref Vector256<float> vs0 = ref Unsafe.Add(ref sourceBase, i);
+                    ref Vector512<float> vs0 = ref Unsafe.Add(ref sourceBase, i);
+                    ref Vector512<float> vd0 = ref Unsafe.Add(ref destinationBase, i);
 
-                    vd0 = Avx.Permute(vs0, control);
-                    Unsafe.Add(ref vd0, 1) = Avx.Permute(Unsafe.Add(ref vs0, 1), control);
-                    Unsafe.Add(ref vd0, 2) = Avx.Permute(Unsafe.Add(ref vs0, 2), control);
-                    Unsafe.Add(ref vd0, 3) = Avx.Permute(Unsafe.Add(ref vs0, 3), control);
+                    vd0 = Vector512Utilities.Shuffle(vs0, control);
+                    Unsafe.Add(ref vd0, (nuint)1) = Vector512Utilities.Shuffle(Unsafe.Add(ref vs0, (nuint)1), control);
+                    Unsafe.Add(ref vd0, (nuint)2) = Vector512Utilities.Shuffle(Unsafe.Add(ref vs0, (nuint)2), control);
+                    Unsafe.Add(ref vd0, (nuint)3) = Vector512Utilities.Shuffle(Unsafe.Add(ref vs0, (nuint)3), control);
```

Comment:
Worth noting that you could do this like the following:
```csharp
Vector512Utilities.Shuffle(Vector512.LoadUnsafe(ref sourceBase), control).StoreUnsafe(ref destinationBase).
Vector512Utilities.Shuffle(Vector512.LoadUnsafe(ref sourceBase, 1), control).StoreUnsafe(ref destinationBase, 1).
Vector512Utilities.Shuffle(Vector512.LoadUnsafe(ref sourceBase, 2), control).StoreUnsafe(ref destinationBase, 2).
Vector512Utilities.Shuffle(Vector512.LoadUnsafe(ref sourceBase, 3), control).StoreUnsafe(ref destinationBase, 3).
```

It shouldn't result in any codegen differences, but avoids needing to manipulate the `byref` and uses direct hardware intrinsic APIs, rather than the Unsafe helpers, so can be easier to read/understand at least IMO.

Reply:
`sourceBase` is `Vector512<float>` so `Vector512.LoadUnsafe(ref sourceBase, i + 0)` would yield `Vector512<Vector512<float>>`. I could use `float` as the base ref type but that complicates the offsetting which confuses me and causes me to make mistakes.